### PR TITLE
Spectral adaptive samplers

### PR DESCRIPTION
**Pull request by vsnever**
This adds the following new samplers:
- `MaskedFrameSampler2D`,
- `SpectralAdaptiveSampler1D`,
- `SpectralAdaptiveSampler2D`,
- `MaskedSpectralAdaptiveSampler2D`.

Also, parameter validation is added for all adaptive samplers (I hope this doesn't break anything).

`SpectralAdaptiveSampler2D` is useful when calculating in a single run multiple filtered camera images with different filters. It has three options for how to reduce the spectral array of normalized errors to a single value for a pixel. I'll be happy to add more options if needed.

There are also two small changes in documentation:
- description of `ratio` parameter of adaptive samplers is added,
- all previously missing pipelines and samplers are added to API reference.

### CODE REVIEW

**Comment by mattngc** [comment]
Wow, had a quick look through and its really nice code Vlad. I think its one Alex would probably want to review.

File: raysect/optical/observer/sampler1d.pyx [hunk 1]
```diff
@@ -153,3 +202,213 @@ cdef class MonoAdaptiveSampler1D(FrameSampler1D):
...
+        elif self.reduction_method == 'mean':
+            for pixel in range(frame.nx):
+                count = 0
+                for sbin in range(frame.ny):
+                    if frame.mean_mv[pixel, sbin] > 0:
+                        normalised_mv[pixel] += error[pixel, sbin] / frame.mean_mv[pixel, sbin]
+                        count += 1
+                if count:
+                    normalised_mv[pixel] /= count
+        else:
```
[the remark this answers is not in the document]
Comment (vsnever) [reply raysect/optical/observer/sampler1d.pyx]: Fixed.

File: raysect/optical/observer/sampler1d.pyx [hunk 2]
```diff
@@ -153,3 +202,213 @@ cdef class MonoAdaptiveSampler1D(FrameSampler1D):
...
+        self._pipeline = value
+
+    @property
+    def fraction(self):
+        return self._fraction
+
+    @fraction.setter
+    def fraction(self, value):
+        if value <= 0 or value > 1.:
+            raise ValueError("fraction must be in the range (0, 1]")
```
[the remark this answers is not in the document]
Comment (vsnever) [reply raysect/optical/observer/sampler1d.pyx]: Fixed.

File: raysect/optical/observer/sampler2d.pyx [hunk 3]
```diff
@@ -58,6 +59,54 @@ cdef class FullFrameSampler2D(FrameSampler2D):
...
+
+    @property
+    def mask(self):
+        return self._mask
+
+    @mask.setter
+    def mask(self, np.ndarray value):
+        if value.ndim != 2:
+            raise ValueError("Mask must be a 2D array")
+        if value.dtype not in (np.int, np.int32, np.int64, np.bool):
```
[the remark this answers is not in the document]
Comment (vsnever) [reply raysect/optical/observer/sampler2d.pyx]: Fixed.

File: raysect/optical/observer/sampler2d.pyx [hunk 4]
```diff
@@ -236,7 +370,428 @@ cdef class MaskedMonoAdaptiveSampler2D(FrameSampler2D):
...
+        if pixels != (frame.nx, frame.ny):
+            raise ValueError('The pixel geometry passed to the frame sampler is inconsistent with the pipeline frame size.')
+
+        min_samples = max(self.min_samples, <int>(frame.samples.max() / self.ratio))
+        error = frame.errors()
+        normalised = np.zeros((frame.nx, frame.ny))
+        normalised_mv = normalised
+
+        # calculated normalised standard error
+        if self.reduction_method == 'weighted':
```
[the remark this answers is not in the document]
Comment (vsnever) [reply raysect/optical/observer/sampler2d.pyx]: I moved them to their own functions.

**Comment by vsnever** [comment]
@CnlPepper, thank you very much for the review. I hope that I fixed all the issues you indicated. Below are some comments.
- The `mask` attribute is added to all 2d samplers. Existing Masked samplers are converted to wrappers and marked as deprecated (they raise respective warning message).
- The setter converts `mask` array to np.bool whatever the original type is. If `mask` is not provided, the all-true mask is created at the first call of `generate_tasks()`. This simplifies the checks. Also, a `memoryview` object is created for the mask (it turns out that numpy boolean arrays are stored in np.uint8).
- Reduction methods of Spectral Adaptive samplers are moved to their own functions. An additional method called 'power_percentile' is added. If `percentile=x`, this method will abort extra sampling when x% of spectral bins **with the highest spectral power** all have normalised errors lower than `cutoff`. This method should be more useful than the simple 'percentile' method.

**Comment by vsnever** [comment]
There was a small formatting error in the docstring of Spectral Adaptive samplers which is fixed now.

File: raysect/optical/observer/sampler1d.pyx [hunk 5]
```diff
@@ -112,25 +163,28 @@ cdef class MonoAdaptiveSampler1D(FrameSampler1D):
         if pixels != frame.length:
             raise ValueError('The pixel geometry passed to the frame sampler is inconsistent with the pipeline frame size.')
 
-        min_samples = max(self.min_samples, frame.samples.max() / self.ratio)
+        min_samples = max(self._min_samples, <int>(frame.samples.max() / self._ratio))
         error = frame.errors()
+        mean = frame.mean_mv  # does memoryview initialisation check before the loop
```
Comment (CnlPepper) [inline raysect/optical/observer/sampler1d.pyx]: Sorry, didn't spot this earlier.

https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#compiler-directives

This change wasn't necessary. The @cython.initializedcheck(False) was missing, if you add this, all the memoryview checks are avoided. Could you please revert this change and add the above decorator. Same goes for the other places.

File: raysect/optical/observer/sampler2d.pyx [hunk 6]
```diff
@@ -27,30 +27,70 @@
...
+
+    @mask.setter
+    def mask(self, np.ndarray value):
+        if value is None:
+            self._mask = None
+        else:
+            if value.ndim != 2:
+                raise ValueError("Mask must be a 2D array.")
+            self._mask = value.astype(np.bool)
+            self._mask_mv = np.frombuffer(self._mask, dtype=np.uint8).reshape(self.mask.shape)
```
Comment (CnlPepper) [inline raysect/optical/observer/sampler2d.pyx]: I've not run this yet, but won't this create a completely separate numpy array, only referenced by the memory view? i.e mask and mask_mv point to different objects? That would be a bad behaviour. Are you sure you can't use bool in the memory view?

Comment (CnlPepper) [reply raysect/optical/observer/sampler2d.pyx]: It might be better to promote the mask to integers internally and convert to bool externally if you can't obtain a bool from the memory view.

Comment (vsnever) [reply raysect/optical/observer/sampler2d.pyx]: For me, this was also a surprise, but numpy stores boolean values not as bits but as bytes (see the description of np.bool [here](https://docs.scipy.org/doc/numpy/user/basics.types.html)), so converting to np.uint8 does not create a new array. You can check it if you want:
```
a = np.ones(4, dtype=bool)
b = np.frombuffer(a, dtype=np.uint8).reshape(2, 2)
a[0] = False
print(b)
```

File: raysect/optical/observer/sampler1d.pyx (hunk 5 above)
Comment (vsnever) [reply raysect/optical/observer/sampler1d.pyx]: It's done.

**Comment by CnlPepper** [comment]
Thanks for the work on this. There are a few tweaks I'll make the the mask interface to make it more user friendly, but I'll leave them until we get closer to release. At present, if you rerun a job with a different pixel size the old mask will be used and error at the frame size check.

**Comment by vsnever** [comment]
Yes, this is a bug. If the mask was not specified by the user, the error should not be raised. Probably the easiest fix is to replace:
```
if pixels != (self._mask.shape[0], self._mask.shape[1]):
    raise ValueError('The pixel geometry passed to the frame sampler is inconsistent with the mask frame size.')
```
with
```
if pixels != (self._mask.shape[0], self._mask.shape[1]):
    if np.all(self._mask):
        self.mask = np.ones(pixels, dtype=np.bool)
    else:
        raise ValueError('The pixel geometry passed to the frame sampler is inconsistent with the mask frame size.')
```
